<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8" />
    <meta name="author" content="Joe Brown" />
    <title>.bed qc</title>
    <link href="https://fonts.googleapis.com/css2?family=Fraunces&display=swap" rel="stylesheet" />
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" rel="stylesheet">

    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js"></script>
    <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
    <script src="https://cdn.plot.ly/plotly-2.2.0.min.js"></script>
    <script src="https://cdn.biowasm.com/v2/aioli/latest/aioli.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/PapaParse/5.3.1/papaparse.min.js"></script>

    <style>
        body {
            height: 100vh;
            margin: 0px;
            padding: 0px;
        }

        footer {
            min-height: 200px;
        }

        .brand {
            font-family: "Fraunces", serif;
            font-size: 2.2rem;
            letter-spacing: -0.05em;
            color: #d91e30 !important;
        }

        .form-control-lg {
            font-size: 2rem;
        }

        .form-control-lg::file-selector-button {
            min-height: 100px;
        }

        label {
            cursor: pointer;
        }

        .info-card {
            min-height: 165px;
        }

        .plot-card {
            min-height: 362px;
        }

        .tab-content>.tab-pane:not(.active),
        .pill-content>.pill-pane:not(.active) {
            display: block;
            height: 0;
            overflow-y: hidden;
        }

        table {
            white-space: pre;
            font-size: 1.2rem;
        }

        .btn-circle {
            width: 23px;
            height: 23px;
            text-align: center;
            padding: 4px 0;
            font-size: 14px;
            line-height: 0.9;
            border-radius: 15px;
            font-weight: bold;
        }
    </style>
</head>

<nav class="navbar">
    <div class="container-fluid row">
        <div class="d-flex flex-wrap align-items-center justify-content-center justify-content-lg-start">
            <a class="navbar-brand brand d-flex align-items-center mb-md-0 me-md-auto text-decoration-none"
                href="https://github.com/quinlan-lab/bedqc">.bed qc</a>
        </div>
        <div class="col-lg-8 mx-auto px-2">
            <label for="file-upload" class="form-label w-100 m-0">
                <div class="col-12 my-3">
                    <h2>Upload .bed</h2>
                </div>
            </label>
            <input class="form-control form-control-lg" type="file" id="file-upload">
            <label for="file-upload" class="form-label w-100">
                <div class="row">
                    <div class="col-12 py-2 text-secondary">
                        Your files and data are never transferred over the internet.
                    </div>
                </div>
            </label>
        </div>
    </div>
</nav>

<div class="offcanvas offcanvas-start" tabindex="-1" id="offcanvas-interval-density"
    aria-labelledby="offcanvas-interval-density-label">
    <div class="offcanvas-header">
        <h5 class="offcanvas-title" id="offcanvas-interval-density-label">Interpreting interval density plots</h5>
        <button type="button" class="btn-close text-reset" data-bs-dismiss="offcanvas" aria-label="Close"></button>
    </div>
    <div class="offcanvas-body">
        <div class="row pb-3">
            <div class="card-body">
                <h3 class="card-subtitle mb-2 text-success">Normal</h3>
                <div class="card-img-top" id="interval-density-normal"></div>
                <p class="card-text pt-2">Based on the known chromosome length and total interval count, the
                    intervals are spread evenly among the chromosomes.
                </p>
            </div>
        </div>
        <div class="row pb-3">
            <div class="card-body">
                <h3 class="card-subtitle mb-2 text-warning">Potential issue</h3>
                <div class="card-img-top" id="interval-density-issue"></div>
                <p class="card-text">Unequal distribution among chromosomes could mean
                    a filter has been applied unevenly or possible file truncation.
                </p>
            </div>
        </div>
    </div>
</div>

<main class="container-fluid" id="content">
    <!-- <label class="form-label">Intersect</label>
    <div class="form-check">
        <input class="form-check-input" type="checkbox" id="intersect-gap">
        <label class="form-check-label" for="intersect-gap">
            gaps
        </label>
    </div> -->

    <!-- Gap intersect -->
    <!-- Our intervals (in this case) do not overlap the centromeres suggesting the selected genome build is not incorrect. -->
    <!-- In most cases, when we overlap our meaningful regions with known gaps we anticipate no centromere overlap. -->
</main>

<div class="container">
    <footer class="d-flex flex-wrap justify-content-between align-items-center py-3 my-4">
      <div class="col-md-4 d-flex align-items-center">
        <!-- <a class="navbar-brand brand d-flex align-items-center mb-md-0 me-md-auto text-decoration-none" -->
                <!-- href="https://github.com/quinlan-lab/bedqc">.bed qc</a> -->
        <!-- <span class="text-muted">&copy; 2021</span> -->
      </div>

      <ul class="nav col-md-4 justify-content-end list-unstyled d-flex">
        <!-- <li class="ms-3"><a class="text-muted" href="https://github.com/quinlan-lab/bedqc"><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-github" viewBox="0 0 16 16"> -->
            <!-- <path d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.012 8.012 0 0 0 16 8c0-4.42-3.58-8-8-8z"/> -->
          <!-- </svg></a></li> -->
      </ul>
    </footer>
  </div>

<script type="module">

    const plot_config = {
        displaylogo: false,
        toImageButtonOptions: {
            format: 'svg', // one of png, svg, jpeg, webp
            filename: 'bedqc',
            height: 600,
            width: 1200,
            scale: 1 // Multiply title/legend/axis/canvas sizes by this factor
        },
        responsive: true,
    }
    const aesthetics = {
        height: 375,
        palette: ["#4e79a7", "#f28e2c", "#e15759", "#76b7b2", "#59a14f", "#edc949", "#af7aa1", "#ff9da7", "#9c755f", "#bab0ab"],
        tickfont: {
            family: "Courier New, monospace",
            size: 16,
            color: "#000",
        },
        labelfont: {
            family: "Raleway, sans-serif",
            size: 12,
        }
    }

    const build_example_plot = ({ x, o, e, divid, ytitle }) => {
        let traces = [{ x: x, y: o, type: "bar", marker: { color: aesthetics.palette[0], line: { color: 'black', width: 1 }, }, name: "Observed" }];
        if (e.length > 0) {
            traces.push({ x: x, y: e, type: "scatter", name: "Expected", mode: "markers", marker: { size: 12, symbol: "circle", color: "#f28e2c", line: { color: 'rgb(255,255,255)', width: 1 }, } })
        }
        let layout = { height: 200, font: aesthetics.labelfont, padding: { t: 0, r: 0, b: 0 }, margin: { t: 0, r: 0, b: 0 }, xaxis: { automargin: true, type: "category", }, yaxis: { title: ytitle, automargin: true, }, legend: { font: { size: 12 } } };
        Plotly.newPlot(divid, traces, layout, { staticPlot: true, responsive: true });
    }

    build_example_plot({
        x: ["chr1", "chr2", "chr3"],
        o: [0.075, 0.079, 0.06],
        e: [0.08, 0.07, 0.06],
        divid: "interval-density-normal",
        ytitle: "Density"
    })

    build_example_plot({
        x: ["chr1", "chr2", "chr3"],
        o: [0.075, 0.13, 0.025],
        e: [0.08, 0.07, 0.06],
        divid: "interval-density-issue",
        ytitle: "Density"
    })

    // build_example_plot({
    //     x: ["chr1", "chr2", "chr3"],
    //     o: [0, 0, 0],
    //     e: [],
    //     divid: "gap-intersect-normal",
    //     ytitle: "Count"
    // })
    // build_example_plot({
    //     x: ["chr1", "chr2", "chr3"],
    //     o: [50, 57, 22],
    //     e: [],
    //     divid: "gap-intersect-issue",
    //     ytitle: "Count"
    // })

    const add_plot_div = (track, title, pb = 2) => {
        let d = document.getElementById("content")

        let exists = document.getElementById(`${track}-row`);
        if (!(exists)) {
            d.insertAdjacentHTML(
                'beforeend',
                `<div class="container-fluid row pb-${pb}">
                    <div class="col-lg-8 mx-auto p-0">
                        <div class="row" id="${track}-row">
                            <div class="col-12">
                                <h5>${title}</h5>
                            </div>
                        </div>
                        <div class="row pt-1" id="${track}-plot-row">
                            <div class="col-12">
                                <div id="${track}-plot"></div>
                            </div>
                        </div>
                    </div>
                </div>
                `
            )
        }
        return Promise.resolve()
    }

    const add_table_row = ({ title, value, warn }) => {
        let divid = title.replace(" ", "-")
        let d = document.getElementById("metadata-table")
        let exists = document.getElementById(divid)
        let text_color = warn ? "text-warn" : "text-primary"
        if (!(exists)) {
            d.insertAdjacentHTML(
                "beforeend",
                `<tr>
                    <td>${title}</td>
                    <td id="${divid}" class="${text_color}"><pre>${value}</pre></td>
                </tr>`
            )
        } else {
            $(`#${divid}`).html(value)
        }
    }

    let CLI = await new Aioli("bedtools/2.29.2");
    const output = await CLI.exec("bedtools --version");
    console.log(`Loaded ${output}`);

    const strip_prefix = (str) => {
        if (str.startsWith("chr")) {
            return str.slice(3)
        }
        return str
    }

    // arr.sort(function (a, b) {
    //     let _a = strip_prefix(a)
    //     let _b = strip_prefix(b)
    //     console.log(_a, _b)
    //     return _a - _b || _a > _b
    // });

    let chromosome_lengths = {};
    let prefix = true

    const to_title = (str) => {
        return str.replace(
            /\w\S*/g,
            (txt) => {
                return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()
            }
        )
    }

    const cli = async (track, chromosomes) => {
        let l = await CLI.ls(".")
        console.log(l)

        // intersect user upload and UCSC track
        let o = await CLI.exec(`bedtools intersect -sorted -a a.bed -b b.bed`)
        o = o.split("\n").map(p => p.split("\t"))

        let chr_count = {}
        o.forEach(row => {
            chr_count[row[0]] = (chr_count[row[0]] || 0) + 1;
        })

        // append a div...
        add_plot_div(track, `${to_title(track)} intersect count`)
        build_bar_plot(
            chromosomes,
            chromosomes.map(chr => chr_count[chr]),
            "Chromosome",
            "Overlap count",
            `${track}-plot`
        )
    }

    const bedtools_intersect = async (genome, track, chromosomes, prefix) => {
        let ucsc_str = ""
        jQuery.getJSON(`https://api.genome.ucsc.edu/getData/track?genome=${genome};track=${track}`, (track_data) => {
            let selection = track_data[track]

            let bed = [];
            chromosomes.forEach(chr => {
                selection[chr].forEach(idx => {
                    if (![prefix]) {
                        bed.push([idx.chrom.substring(3), idx.chromStart, idx.chromEnd])
                    } else {
                        bed.push([idx.chrom, idx.chromStart, idx.chromEnd])
                    }
                })
            })
            // sort -k1,1 -k2,2n
            bed.sort(function (a, b) {
                return a[1] > b[1];
            });
            bed.sort(function (a, b) {
                return a[0] > b[0];
            });
            // toString...
            let str = "";
            bed.map(a => str += a.join("\t") + "\n");
            CLI.fs.writeFile("b.bed", str);
        }).then(() => {
            cli(track, chromosomes)
        })
    }

    // const checkbox_change = (event) => {
    //     if (event.target.checked == false) {
    //         // remove the div of the unchecked intersection
    //         // intersect-gap and need to remove gap-row, gap-plot-row
    //         let pfx = event.target.id.split("-")[1]
    //         document.getElementById(`${pfx}-row`).remove()
    //         document.getElementById(`${pfx}-plot-row`).remove()
    //     } else {
    //         let file = $('#file-upload').prop('files')[0]
    //         let genome = $('#genome-select').val()
    //         if ($("#intersect-gap").prop("checked") == true) {
    //             bedtools_intersect(genome, "gap", m.chromosomes, m.chr_prefix)
    //         }
    //         // additional checkboxes
    //     }
    // }

    // const handle_checkboxes = ({ file, genome, chroms, chr_prefix }) => {
    //     if ($("#intersect-gap").prop("checked") == true) {
    //         bedtools_intersect(genome, "gap", chroms, chr_prefix)
    //     }
    //     // additional checkboxes
    // }

    const delimiter_string = (sep) => {
        if (sep == "\t") {
            return "\\t"
        } else {
            return sep
        }
    }

    const newline_string = (nl) => {
        if (nl == "\n") {
            return "\\n"
        } else if (nl == "\r\n") {
            return "\\r\\n"
        } else {
            return `${nl}`
        }
    }

    const genome_build = () => {
        let d = document.getElementById("content")

        d.insertAdjacentHTML(
            "beforeend",
            `<div class="container-fluid row pb-2">
                <div class="col-lg-8 mx-auto p-0">
                    <div class="col-12 my-3" id="genome-build-anchor">
                        <h2>Genome build</h2>
                    </div>
                    <div class="row">
                        <div class="col-3">
                            <select class="form-select form-select-lg" id="species-select" disabled>
                                <option selected>human</option>
                            </select>
                        </div>
                        <div class="col-6">
                            <select class="form-select form-select-lg" id="genome-select">
                                <option selected>hg38</option>
                                <option>hg19</option>
                            </select>
                        </div>
                        <div class="col-3">
                            <div class="d-grid gap-2">
                                <button type="button" class="btn btn-primary btn-lg position-relative" id="genome-select-action">
                                    Apply
                                    <span id="button-alert" class="position-absolute top-0 start-100 translate-middle p-2 bg-danger border border-light rounded-circle" hidden>
                                        <span class="visually-hidden">New alerts</span>
                                    </span>
                                </button>
                            </div>
                        </div>
                    </div>
                    <div class="row">
                        <div class="col-12 py-2 text-secondary">
                            Select the genome build of the BED file. This option can be changed later to test if other genome builds better fit these data.
                        </div>
                    </div>
                </div>
            </div>
            `
        )
        document.getElementById("genome-select-action").addEventListener("click", genome_selected);
        document.getElementById("genome-select").addEventListener("change", update_button);
        return Promise.resolve()
    }

    const genome_selected = () => {
        // button was clicked, show user a check that we're done
        // $("#genome-select-action").prop("disabled", true)
        $("#button-alert").prop("hidden", true)

        let genome = $('#genome-select').val()
        let chromosome_lengths;
        jQuery.getJSON(`https://api.genome.ucsc.edu/list/chromosomes?genome=${genome}`, (d) => {
            chromosome_lengths = d.chromosomes;
            // adds non-chr prefixed chroms in addition to chr prefixed result
            Object.keys(chromosome_lengths).map(k => chromosome_lengths[k.substring(3)] = chromosome_lengths[k]);
        }).then(() => {
            let genome_length = 0;
            global_data.chromosomes.map(x => genome_length += chromosome_lengths[x]);

            // add_plot_div('interval-density', `Interval density across ${genome}`, 5)

            let d = document.getElementById("content")
            let exists = document.getElementById(`interval-density-row`);

            // if it exists, update the title and ALL subsequent plots
            if (exists) {
                // update the title
                $("#interval-density-plot-title").html(`Interval density across ${genome}`)
                // update ALL plots...
            } else {
                d.insertAdjacentHTML(
                    'beforeend',
                    `<div class="container-fluid row pb-5">
                        <div class="col-lg-8 mx-auto p-0">
                            <div class="d-flex flex-row" id="interval-density-row">
                                <div class="pe-2">
                                    <h5 id="interval-density-plot-title">Interval density across ${genome}</h5>
                                </div>
                                <a class="btn btn-primary btn-circle" data-bs-toggle="offcanvas" href="#offcanvas-interval-density" role="button" aria-controls="offcanvas-interval-density">?</a>
                            </div>
                            <div class="row pt-1" id="interval-density-plot-row">
                                <div class="col-12">
                                    <div id="interval-density-plot"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                    `
                )
            }

            build_bar_plot(
                global_data.chromosomes,
                global_data.chromosomes.map(x => global_data.chr_interval_count[x] / global_data.interval_count_filtered),
                "Chromosome",
                "Density",
                "interval-density-plot",
                {
                    x: global_data.chromosomes,
                    y: global_data.chromosomes.map(x => (chromosome_lengths[x] / genome_length)),
                    type: "scatter",
                    name: "Expected",
                    mode: "markers",
                    marker: {
                        size: 12, symbol: "circle", color: "#f28e2c", line: {
                            color: 'rgb(255,255,255)',
                            width: 1
                        },
                    }
                }
            );

            document.getElementById("genome-build-anchor").scrollIntoView({behavior: "smooth", block: "start"});

            build_hypothesis_test()
        })
    }

    const build_hypothesis_test = () => {
        global_data.hypothesis_test_count += 1

        // Hypothesis test #1
        // Intersect [Gap] from [hg19]
        // Subtitle: If your intervals are coding regions and intersect with many gaps across chromosomes, [hg19] may not be the
        // underlying genome build of [bedfile name].

        let d = document.getElementById("content")
        d.insertAdjacentHTML(
            "beforeend",
            `<div class="container-fluid row pb-2">
                <div class="col-lg-8 mx-auto p-0">
                    <label for="genome-select" class="form-label w-100 m-0">
                        <div class="col-12 my-3">
                            <h2>Hypothesis test &#35;${global_data.hypothesis_test_count}</h2>
                        </div>
                    </label>
                    <div class="row">
                        <div class="col-3">
                            <select class="form-select form-select-lg" id="species-select" disabled>
                                <option selected>human</option>
                            </select>
                        </div>
                        <div class="col-6">
                            <select class="form-select form-select-lg" id="genome-select">
                                <option selected>hg38</option>
                                <option>hg19</option>
                            </select>
                        </div>
                        <div class="col-3">
                            <div class="d-grid gap-2">
                                <button type="button" class="btn btn-primary btn-lg position-relative" id="genome-select-action">
                                    Apply
                                    <span id="button-alert" class="position-absolute top-0 start-100 translate-middle p-2 bg-danger border border-light rounded-circle" hidden>
                                        <span class="visually-hidden">New alerts</span>
                                    </span>
                                </button>
                            </div>
                        </div>
                    </div>
                    <label for="genome-select" class="form-label w-100">
                        <div class="row">
                            <div class="col-12 py-2 text-secondary">
                                Select the genome build of the BED file. This option can be changed later to test if other genome builds better fit these data.
                            </div>
                        </div>
                    </label>
                </div>
            </div>
            `
        )
        bedtools_intersect("hg38", "gap", global_data.chromosomes, global_data.chr_prefix)
    }

    const update_button = (event) => {
        // $("#genome-select-action").prop("disabled", false)
        // show the user that they have changed the genome build and that we have to re-build the plots
        $("#button-alert").prop("hidden", false)
    }

    let global_data = {
        hypothesis_test_count: 0
    };

    const file_upload = (event) => {
        // clean up existing divs
        $("#content").empty();

        let d = document.getElementById("content")
        d.insertAdjacentHTML(
            "beforeend",
            `<div class="container-fluid row pb-3">
                <div class="col-lg-8 mx-auto bg-light border" id="file-validation" hidden>
                    <table class="table table-borderless m-0">
                        <tbody id="metadata-table">
                        </tbody>
                    </table>
                </div>
            </div>`
        )

        let bed = [];

        let data = {
            newline: "\n",
            delimiter: "\t",
            n_of_fields: [],
            missing_data: 0,
            interval_count: 0,
            interval_length_total: 0,
            interval_lengths: [],
            chr_interval_length: {},
            filtered_intervals: 0,
            zero_length_intervals: 0,
            extends_beyond_chr: 0,
            chr_prefix: true,
            chromosomes: [],
            all_chromosomes: []
        };
        global_data = {
            chr_interval_count: {},
            interval_count_filtered: 0,
            chr_prefix: true,
            chromosomes: []
        }
        if ($('#file-upload').prop('files')[0]) {
            Papa.parse($('#file-upload').prop('files')[0], {
                skipEmptyLines: true,
                header: false,
                delimiter: "\t",
                step: (res, parser) => {
                    data.interval_count += 1;

                    // incorrect line ending
                    if (!(res.meta.linebreak == "\n")) {
                        console.log("aborting due to bad line break")
                        data.newline = res.meta.linebreak
                        parser.abort()
                    }
                    // not tab delimited
                    if (!(res.meta.delimiter == "\t")) {
                        console.log("aborting due to bad delimiter")
                        data.delimiter = res.meta.delimiter
                        parser.abort()
                    }
                    // most likely a delimiter issue
                    if (res.data.length < 3) {
                        // this file does not have at least 3 fields per line
                        // offending line is d.interval_count with values
                        // res.data
                        console.log("aborting:", res.data)
                        data.missing_data = data.interval_count
                        parser.abort()
                    }
                    // collect shape of bed data
                    if (!(data.n_of_fields.includes(res.data.length))) {
                        data.n_of_fields.push(res.data.length)
                    }

                    let chr = res.data[0];
                    let current_length = res.data[2] - res.data[1];
                    // console.log("chr", chr)
                    if (!(chr.startsWith('chr'))) {
                        global_data.chr_prefix = false
                    }

                    // set the chromosome order by appearance in bed file
                    if (!(chr in data.chr_interval_length)) {
                        data.all_chromosomes.push(chr)
                    }

                    if (chr.match(/fix|random|alt|Un|hap/g) === null) {
                        global_data.interval_count_filtered += 1;
                        bed.push(res.data.slice(0, 3))
                        // this eval is duplicated
                        if (!(chr in data.chr_interval_length)) {
                            global_data.chromosomes.push(chr);
                        }
                    }

                    // malformatted entry; 0-based start, 1-based end
                    if (current_length == 0) {
                        data.zero_length_intervals += 1;
                    }

                    data.interval_length_total += current_length;
                    data.interval_lengths.push(current_length);
                    data.chr_interval_length[chr] = (data.chr_interval_length[chr] || 0) + current_length;
                    global_data.chr_interval_count[chr] = (global_data.chr_interval_count[chr] || 0) + 1;
                },
                complete: (res, file) => {
                    // resize the upload area
                    let el = document.getElementById("file-upload")
                    // $(".form-control-lg::file-selector-button").attr("style", "min-height:")
                    // show the file stats
                    $("#file-validation").prop("hidden", false)
                    if (!(data.delimiter == "\t")) {
                        add_table_row({ title: "File delimiter", value: delimiter_string(data.delimiter), warn: true })
                    } else {
                        add_table_row({ title: "File delimiter", value: delimiter_string(data.delimiter) })
                    }
                    if (!(data.newline == "\n")) {
                        add_table_row({ title: "File newline", value: newline_string(data.newline), warn: true })
                    } else {
                        add_table_row({ title: "File newline", value: newline_string(data.newline) })
                    }
                    if (!(data.missing_data == 0)) {
                        add_table_row({ title: "Incomplete data on line", value: data.missing_data, warn: true })
                        return
                    }
                    if (data.newline == "\n" && data.delimiter == "\t") {
                        // < !--number of fields per line-- >
                        add_table_row({ title: "BED fields", value: data.n_of_fields.join(",") })
                        // < !--interval count-- >
                        add_table_row({ title: "Total interval count", value: data.interval_count.toLocaleString() })
                        // < !--total nucleotides-- >
                        add_table_row({ title: "Total nucleotides", value: data.interval_length_total.toLocaleString() })
                        // < !--0 - length intervals-- >
                        add_table_row({ title: "0-length intervals", value: data.zero_length_intervals.toLocaleString() })
                    }
                    // show the interval size dist
                    add_plot_div('interval-size-dist', 'Interval size distribution', 5).then(() => {
                        var trace = {
                            x: data.interval_lengths,
                            type: 'histogram',
                            marker: { color: aesthetics.palette[0], },
                        };
                        let layout = {
                            height: aesthetics.height,
                            font: aesthetics.labelfont,
                            padding: { t: 0, r: 0, },
                            margin: { t: 0, r: 0, },
                            xaxis: { title: "Interval length", automargin: true, },
                            yaxis: { title: "Count", automargin: true, },
                            legend: { font: { size: 16 } }
                        };
                        Plotly.newPlot('interval-size-dist-plot', [trace], layout, plot_config)
                    })
                    // sort by start, asc
                    bed.sort(function (a, b) {
                        return parseInt(a[1]) > parseInt(b[1]);
                    });
                    // sort by chrom, asc
                    bed.sort(function (a, b) {
                        return a[0] > b[0];
                    });
                    genome_build();
                    // convert bed to string
                    let str = "";
                    bed.map(a => str += a.join("\t") + "\n");
                    CLI.fs.writeFile("a.bed", str);
                }
            })
        }
    }

    const onChange = (event) => {
        document.getElementById("intersect-gap").addEventListener('change', checkbox_change);

        let genome = $('#genome-select').val()
        let chromosome_lengths;
        jQuery.getJSON(`https://api.genome.ucsc.edu/list/chromosomes?genome=${genome}`, (d) => {
            chromosome_lengths = d.chromosomes
            // adds non-chr prefixed chroms in addition to chr prefixed result
            Object.keys(chromosome_lengths).map(k => chromosome_lengths[k.substring(3)] = chromosome_lengths[k])
        }).then(() => {
            let interval_length = {}
            let interval_count = {}
            // new upload
            m = {
                interval_count: 0,
                interval_count_filtered: 0,
                interval_length_total: 0,
                interval_lengths: [],
                filtered_intervals: 0,
                zero_length_intervals: 0,
                extends_beyond_chr: 0,
                chr_prefix: true,
                chromosomes: [],
                all_chromosomes: []
            }
            if ($('#file-upload').prop('files')[0]) {
                Papa.parse($('#file-upload').prop('files')[0], {
                    delimiter: "\t",
                    skipEmptyLines: false,
                    newline: "\n",
                    header: false,
                    step: (results, parser) => {
                        m.interval_count += 1;
                        // check number of toks testing the delimiter
                        // likely going to need lengths in the future
                        let chr = results.data[0];
                        // skip unhelpful chromosomes
                        if (chr.match(/fix|random|alt|Un|hap/g) === null) {
                            m.interval_count_filtered += 1;
                            if (results.data.length < 3) {
                                // invalid format or separator
                                // raise format error
                                results.data = []
                                console.log("invalid format (chrom, start, stop) or separator ('\\t')")
                                return
                            }

                            // check for 'chr' prefix
                            if (!(chr.startsWith('chr'))) {
                                m.chr_prefix = false
                                // will need to check if some have chr and some do not
                            }

                            // set the chromosome order by appearance in bed file
                            if (!(chr in interval_length)) {
                                m.chromosomes.push(chr);
                            }

                            // malformatted entry; 0-based start, 1-based end
                            let current_length = results.data[2] - results.data[1]
                            if (current_length == 0) {
                                m.zero_length_intervals += 1
                            }

                            m.interval_length_total += current_length

                            // check if end extends beyond chrom end
                            // this could also be a plot

                            m.interval_lengths.push(current_length)
                            interval_length[chr] = (interval_length[chr] || 0) + current_length;
                            interval_count[chr] = (interval_count[chr] || 0) + 1;
                        }
                        else {
                            m.filtered_intervals += 1
                        }
                    },
                    complete: (results, file) => {

                        var element = document.getElementById("results-tab");
                        element.classList.remove("disabled");

                        var someTabTriggerEl = document.querySelector('#results-tab')
                        var tab = new bootstrap.Tab(someTabTriggerEl)
                        tab.show()

                        $("#summary-filename").html(file.name)
                        $("#summary-interval-count").html(m.interval_count.toLocaleString())
                        $("#summary-total-bp").html(m.interval_length_total.toLocaleString())
                        if (m.filtered_intervals > 0) {
                            add_table_row({
                                title: "Filtered intervals",
                                value: m.filtered_intervals.toLocaleString()
                            })
                        }
                        if (m.zero_length_intervals > 0) {
                            add_table_row({
                                title: "0-length intervals",
                                value: m.zero_length_intervals.toLocaleString()
                            })
                        }

                        add_plot_div('interval-density', 'Interval density')
                        // length corrected for only displayed chromosomes
                        let genome_length = 0
                        m.chromosomes.map(x => genome_length += chromosome_lengths[x])

                        build_bar_plot(
                            m.chromosomes,
                            m.chromosomes.map(x => interval_count[x] / m.interval_count_filtered),
                            "Chromosome",
                            "Density",
                            "interval-density-plot",
                            {
                                x: m.chromosomes,
                                y: m.chromosomes.map(x => (chromosome_lengths[x] / genome_length)),
                                type: "scatter",
                                name: "Expected",
                                mode: "markers",
                                marker: {
                                    size: 12, symbol: "circle", color: "#f28e2c", line: {
                                        color: 'rgb(255,255,255)',
                                        width: 1
                                    },
                                }
                            }
                        );
                        add_plot_div('interval-size-dist', 'Interval size distribution').then(() => {
                            var trace = {
                                x: m.interval_lengths,
                                type: 'histogram',
                                marker: {
                                    color: aesthetics.palette[0],
                                },
                            };
                            let layout = {
                                height: aesthetics.height,
                                font: aesthetics.labelfont,
                                padding: {
                                    t: 0,
                                    r: 0,
                                },
                                margin: {
                                    t: 0,
                                    r: 0,
                                },
                                xaxis: {
                                    title: "Interval length",
                                    automargin: true,
                                },
                                yaxis: {
                                    title: "Count",
                                    automargin: true,
                                },
                                legend: {
                                    font: {
                                        size: 16
                                    }
                                }
                            };
                            Plotly.newPlot('interval-size-dist-plot', [trace], layout, plot_config)
                        })

                        // handle checkboxes
                        handle_checkboxes({
                            file: file,
                            genome: genome,
                            chroms: m.chromosomes,
                            chr_prefix: m.chr_prefix
                        })
                    }
                })
            }
        })
    }

    // const mean = (data) => {
    //     let result = 0;
    //     let n = 0;
    //     for (let i = 0; i < data.length; i++) {
    //         let d = data[i];
    //         if (d >= 0) {
    //             n += 1;
    //             result += d;
    //         }
    //     }
    //     return result / n;
    // };

    const build_bar_plot = (chrs, o, x_title, y_title, div, trace = false) => {
        let traces = [
            {
                x: chrs,
                y: o,
                type: "bar",
                marker: {
                    color: aesthetics.palette[0],
                    line: {
                        color: 'black',
                        width: 1
                    },
                },
                name: "Observed"
            }
        ];

        if (trace) {
            traces.push(trace)
        }
        // console.log(traces)

        let layout = {
            height: aesthetics.height,
            font: aesthetics.labelfont,
            padding: {
                t: 0,
                r: 0,
            },
            margin: {
                t: 0,
                r: 0,
            },
            xaxis: {
                title: x_title,
                automargin: true,
                type: "category",
            },
            yaxis: {
                title: y_title,
                automargin: true,
            },
            legend: {
                font: {
                    size: 16
                }
            }
        };

        return Plotly.newPlot(div, traces, layout, plot_config);
    };

    $(document).ready(function () {
        document.getElementById("file-upload").addEventListener('change', file_upload);
        // document.getElementById("genome-select").addEventListener('change', onChange);
    });
</script>

</html>
