<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8" />
    <meta name="author" content="Joe Brown" />
    <title>.bed qc</title>
    <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Fraunces&display=swap">
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons@1.7.2/font/bootstrap-icons.css">

    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js"></script>
    <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
    <script src="https://cdn.plot.ly/plotly-2.2.0.min.js"></script>
    <script src="https://cdn.biowasm.com/v2/aioli/latest/aioli.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/PapaParse/5.3.1/papaparse.min.js"></script>

    <style>
        body {
            height: 100vh;
            margin: 0px;
            padding: 0px;
        }

        footer {
            min-height: 200px;
        }

        .brand {
            font-family: "Fraunces", serif;
            font-size: 2.2rem;
            letter-spacing: -0.05em;
            color: #d91e30 !important;
        }

        .form-control-lg {
            font-size: 2rem;
        }

        .form-control-lg::file-selector-button {
            min-height: 100px;
        }

        label {
            cursor: pointer;
        }

        .info-card {
            min-height: 165px;
        }

        .plot-card {
            min-height: 362px;
        }

        .tab-content>.tab-pane:not(.active),
        .pill-content>.pill-pane:not(.active) {
            display: block;
            height: 0;
            overflow-y: hidden;
        }

        table {
            white-space: pre;
            font-size: 1.2rem;
        }

        .btn-circle {
            width: 18px;
            height: 18px;
            text-align: center;
            padding: 4px 0;
            font-size: 14px;
            line-height: 0.5;
            border-radius: 15px;
            font-weight: bold;
        }

        .text-primary code {
            color: rgba(var(--bs-primary-rgb), var(--bs-text-opacity)) !important
        }

        .btn-floating {
            line-height: 1.95;
            font-size: 1.4rem;
        }
    </style>
</head>

<nav class="navbar">
    <div class="container-fluid row">
        <div class="d-flex flex-wrap align-items-center justify-content-center justify-content-lg-start">
            <a class="navbar-brand brand d-flex align-items-center mb-md-0 me-md-auto text-decoration-none"
                href="https://github.com/quinlan-lab/bedqc">.bed qc</a>
        </div>
        <div class="col-lg-8 mx-auto px-2">
            <label for="file-upload" class="form-label w-100 m-0">
                <div class="col-12 my-3">
                    <h2>Upload .bed</h2>
                </div>
            </label>
            <input class="form-control form-control-lg" type="file" id="file-upload">
            <label for="file-upload" class="form-label w-100">
                <div class="row">
                    <div class="col-12 py-2 text-secondary">
                        Your files and data are never transferred over the internet.
                    </div>
                </div>
            </label>
        </div>
        <div class="col-lg-8 mx-auto px-2 py-5" id="introduction">
            <p>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean sit amet tortor erat. Etiam sed sapien
                convallis, tempor lorem id, gravida erat. Mauris mattis dictum neque sed porta. Duis non nibh sapien.
                Morbi fermentum lacinia tortor a bibendum. Suspendisse ipsum felis, hendrerit eget lacinia at, porta nec
                est. Pellentesque quis finibus dui. Ut varius condimentum elementum. Sed vel arcu leo. Fusce ultrices at
                diam in ullamcorper.
            </p>
            <p>
                Aliquam at diam id risus tempor consequat eu eget est. Ut posuere tincidunt vestibulum. Mauris nec
                turpis at massa euismod rutrum a ut lorem. Aliquam pharetra condimentum accumsan. Sed quis nibh quis mi
                tristique tristique id a eros. Nulla egestas aliquam fermentum. Etiam ut aliquam mauris, et elementum
                diam. Integer eu convallis sapien. Curabitur ante erat, lobortis ac dui sit amet, iaculis faucibus
                magna. Proin a orci orci.
            </p>
            <p>
                Link to the BED specification.
            </p>
            <p>
                This tool makes use of the <a href="https://github.com/biowasm/aioli" class="link-dark">biowasm
                    project</a>.
            </p>
        </div>
    </div>
</nav>

<div class="offcanvas offcanvas-start" tabindex="-1" id="offcanvas-interval-density"
    aria-labelledby="offcanvas-interval-density-label">
    <div class="offcanvas-header">
        <h5 class="offcanvas-title" id="offcanvas-interval-density-label">Interpreting interval density plots</h5>
        <button type="button" class="btn-close text-reset" data-bs-dismiss="offcanvas" aria-label="Close"></button>
    </div>
    <div class="offcanvas-body">
        <div class="row pb-3">
            <div class="card-body">
                <h3 class="card-subtitle mb-2 text-success">Normal</h3>
                <div class="card-img-top" id="interval-density-normal"></div>
                <p class="card-text pt-2">Based on the known chromosome length and total interval count, the
                    intervals are spread evenly among the chromosomes.
                </p>
            </div>
        </div>
        <div class="row pb-3">
            <div class="card-body">
                <h3 class="card-subtitle mb-2 text-warning">Potential issue</h3>
                <div class="card-img-top" id="interval-density-issue"></div>
                <p class="card-text">Unequal distribution among chromosomes could mean
                    a filter has been applied unevenly or possible file truncation.
                </p>
            </div>
        </div>
    </div>
</div>

<div class="offcanvas offcanvas-start" tabindex="-1" id="offcanvas-interval-counts"
    aria-labelledby="offcanvas-interval-counts-label">
    <div class="offcanvas-header">
        <h5 class="offcanvas-title" id="offcanvas-interval-counts-label">Interpreting interval count plots</h5>
        <button type="button" class="btn-close text-reset" data-bs-dismiss="offcanvas" aria-label="Close"></button>
    </div>
    <div class="offcanvas-body">
        <div class="row pb-3">
            <div class="card-body">
                <h3 class="card-subtitle mb-2 text-success">Normal</h3>
                <div class="card-img-top" id="interval-counts-normal"></div>
                <p class="card-text pt-2">Based on the known chromosome length and total interval count, the
                    intervals are spread evenly among the chromosomes.
                </p>
            </div>
        </div>
        <div class="row pb-3">
            <div class="card-body">
                <h3 class="card-subtitle mb-2 text-warning">Potential issue</h3>
                <div class="card-img-top" id="interval-counts-issue"></div>
                <p class="card-text">Unequal distribution among chromosomes could mean
                    a filter has been applied unevenly or possible file truncation.
                </p>
            </div>
        </div>
    </div>
</div>

<div class="offcanvas offcanvas-start" tabindex="-1" id="offcanvas-Intersect-gap"
    aria-labelledby="offcanvas-intersect-gap-label">
    <div class="offcanvas-header">
        <h5 class="offcanvas-title" id="offcanvas-intersect-gap-label">Interpreting gap intersection plots</h5>
        <button type="button" class="btn-close text-reset" data-bs-dismiss="offcanvas" aria-label="Close"></button>
    </div>
    <div class="offcanvas-body">
        <div class="row pb-3">
            <div class="card-body">
                <h3 class="card-subtitle mb-2 text-success">Normal</h3>
                <div class="card-img-top" id="intersect-gap-normal"></div>
                <p class="card-text pt-2">Our intervals (in this case) do not overlap the centromeres
                    suggesting the selected genome build is <span class="fw-bold fst-italic">not
                        incorrect</span>.
                </p>
            </div>
        </div>
        <div class="row pb-3">
            <div class="card-body">
                <h3 class="card-subtitle mb-2 text-warning">Potential issue</h3>
                <div class="card-img-top" id="intersect-gap-issue"></div>
                <p class="card-text">In most cases, when we overlap our meaningful regions
                    with known gaps we anticipate no centromere overlap.
                </p>
            </div>
        </div>
    </div>
</div>

<main class="container-fluid" id="content">
</main>

<div class="container">
    <footer class="d-flex flex-wrap justify-content-between align-items-center py-3 my-4">
        <div class="col-md-4 d-flex align-items-center">
            <!-- <a class="navbar-brand brand d-flex align-items-center mb-md-0 me-md-auto text-decoration-none" -->
            <!-- href="https://github.com/quinlan-lab/bedqc">.bed qc</a> -->
            <!-- <span class="text-muted">&copy; 2021</span> -->
        </div>

        <!-- <ul class="nav col-md-4 justify-content-end list-unstyled d-flex"> -->
        <!-- <li class="ms-3"><a class="text-muted" href="https://github.com/quinlan-lab/bedqc"><svg -->
        <!-- xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" -->
        <!-- class="bi bi-github" viewBox="0 0 16 16"> -->
        <!-- <path -->
        <!-- d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.012 8.012 0 0 0 16 8c0-4.42-3.58-8-8-8z" /> -->
        <!-- </svg></a></li> -->
        <!-- </ul> -->
    </footer>
</div>

<script type="module">

    const plot_config = {
        displaylogo: false,
        toImageButtonOptions: {
            format: 'svg', // one of png, svg, jpeg, webp
            filename: 'bedqc',
            height: 600,
            width: 1200,
            scale: 1 // Multiply title/legend/axis/canvas sizes by this factor
        },
        responsive: true,
    }
    const aesthetics = {
        height: 375,
        palette: ["#4e79a7", "#f28e2c", "#e15759", "#76b7b2", "#59a14f", "#edc949", "#af7aa1", "#ff9da7", "#9c755f", "#bab0ab"],
        tickfont: {
            family: "Courier New, monospace",
            size: 16,
            color: "#000",
        },
        labelfont: {
            family: "Raleway, sans-serif",
            size: 12,
        }
    }

    const sort_alphanum = (a, b) => a.localeCompare(b, 'en', { numeric: true })

    const build_example_plot = ({ x, o, e, divid, ytitle }) => {
        let traces = [{ x: x, y: o, type: "bar", marker: { color: aesthetics.palette[0], line: { color: 'black', width: 1 }, }, name: "Observed" }];
        if (e.length > 0) {
            traces.push({ x: x, y: e, type: "scatter", name: "Expected", mode: "markers", marker: { size: 12, symbol: "circle", color: "#f28e2c", line: { color: 'rgb(255,255,255)', width: 1 }, } })
        }
        let layout = { height: 200, font: aesthetics.labelfont, padding: { t: 0, r: 0, b: 0 }, margin: { t: 0, r: 0, b: 0 }, xaxis: { automargin: true, type: "category", }, yaxis: { title: ytitle, automargin: true, }, legend: { font: { size: 12 } } };
        Plotly.newPlot(divid, traces, layout, { staticPlot: true, responsive: true });
    }

    build_example_plot({
        x: ["chr1", "chr2", "chr3"],
        o: [0.075, 0.079, 0.06],
        e: [0.08, 0.07, 0.06],
        divid: "interval-density-normal",
        ytitle: "Density"
    })

    build_example_plot({
        x: ["chr1", "chr2", "chr3"],
        o: [0.075, 0.13, 0.025],
        e: [0.08, 0.07, 0.06],
        divid: "interval-density-issue",
        ytitle: "Density"
    })

    build_example_plot({
        x: ["chr1", "chr2", "chr3"],
        o: [100, 90, 80],
        e: [95, 94, 78],
        divid: "interval-counts-normal",
        ytitle: "Density"
    })

    build_example_plot({
        x: ["chr1", "chr2", "chr3"],
        o: [100, 132, 18],
        e: [95, 94, 78],
        divid: "interval-counts-issue",
        ytitle: "Density"
    })

    build_example_plot({
        x: ["chr1", "chr2", "chr3"],
        o: [0, 0, 0],
        e: [],
        divid: "intersect-gap-normal",
        ytitle: "Count"
    })

    build_example_plot({
        x: ["chr1", "chr2", "chr3"],
        o: [50, 57, 22],
        e: [],
        divid: "intersect-gap-issue",
        ytitle: "Count"
    })

    const add_plot_div = (track, title, pb = 2) => {
        let d = document.getElementById("content")

        let exists = document.getElementById(`${track}-row`);
        if (!(exists)) {
            d.insertAdjacentHTML(
                'beforeend',
                `<div class="container-fluid row pb-${pb}">
                    <div class="col-lg-8 mx-auto p-0">
                        <div class="row" id="${track}-row">
                            <div class="col-12">
                                <h5>${title}</h5>
                            </div>
                        </div>
                        <div class="row pt-1" id="${track}-plot-row">
                            <div class="col-12">
                                <div id="${track}-plot"></div>
                            </div>
                        </div>
                    </div>
                </div>
                `
            )
        }
        return Promise.resolve()
    }

    const add_table_row = ({ title, value, warn, validation }) => {
        let divid = title.replace(" ", "-")
        let d = document.getElementById("metadata-table")
        let exists = document.getElementById(divid)
        let text_color = warn ? "text-danger" : "text-primary"
        let validation_col = ""
        if (validation) {
            validation_col = warn ? `<td class="text-danger"><i class="bi bi-exclamation-triangle-fill"></i></td>` : `<td class="text-success"><i class="bi bi-check-circle-fill"></i></td>`
        }
        if (!(exists)) {
            d.insertAdjacentHTML(
                "beforeend",
                `<tr>
                    <td>${title}</td>
                    <td id="${divid}" class="${text_color}"><pre>${value}</pre></td>
                    ${validation_col}
                </tr>`
            )
        } else {
            $(`#${divid}`).html(value)
        }
    }

    let CLI = await new Aioli("bedtools/2.29.2");
    const output = await CLI.exec("bedtools --version");
    console.log(`Loaded ${output}`);

    const strip_prefix = (str) => {
        if (str.startsWith("chr")) {
            return str.slice(3)
        }
        return str
    }

    let chromosome_lengths = {};
    let prefix = true

    const to_title = (str) => {
        return str.replace(
            /\w\S*/g,
            (txt) => {
                return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()
            }
        )
    }

    const scroll_to = (id) => {
        const yOffset = -10;
        const element = document.getElementById(id);
        const y = element.getBoundingClientRect().top + window.pageYOffset + yOffset;

        window.scrollTo({ top: y, behavior: 'smooth' });
    }

    const bedtools_intersect = async (genome, track, chromosomes, prefix) => {
        let remote_file = `bedqc.s3.amazonaws.com/${genome}_${track}.bed`
        if (!(global_data.chr_prefix)) {
            remote_file = `bedqc.s3.amazonaws.com/${genome}_${track}_noprefix.bed`
        }
        try {
            await CLI.mount(`https://${remote_file}`)
        } catch (error) {
            console.log("file likely already exists. ignoring error.")
            console.error(error);
        }
        // console.log("ls:", await CLI.ls("."));
        // intersect user upload and UCSC track
        let o = await CLI.exec(`bedtools intersect -sorted -a a.bed -b ${remote_file.replace("/", "-")}`)
        o = o.split("\n").map(p => p.split("\t"))

        let chr_count = {}
        o.forEach(row => {
            chr_count[row[0]] = (chr_count[row[0]] || 0) + 1;
        })

        let d = document.getElementById("content");
        let idx = global_data.input_test_count;
        let primary = $(`#hypothesis-primary-${idx}`).val();
        let secondary = $(`#hypothesis-secondary-${idx}`).val();

        genome = $(`#hypothesis-genome-${idx}`).val();

        d.insertAdjacentHTML(
            'beforeend',
            `<div class="container-fluid row pb-3">
                <div class="d-flex col-1 justify-content-end"><code>Out[${idx}]</code></div>
                <div class="col-lg-11 p-0">
                    <div class="d-flex flex-row">
                        <div class="pe-2">
                            <h6>${to_title(primary)} ${genome} ${secondary} count</h6>
                        </div>
                        <a class="btn btn-primary btn-circle" data-bs-toggle="offcanvas" href="#offcanvas-${primary}-${secondary}" role="button" aria-controls="offcanvas-${primary}-${secondary}">?</a>
                    </div>
                    <div class="row pt-1">
                        <div class="col-12">
                            <div id="hypothesis-${idx}-plot"></div>
                        </div>
                    </div>
                </div>
            </div>
            `
        )

        build_bar_plot(
            chromosomes,
            chromosomes.map(chr => chr_count[chr]),
            "Chromosome",
            "Overlap count",
            `hypothesis-${idx}-plot`
        )
    }

    const delimiter_string = (sep) => {
        if (sep == "\t") {
            return "tab"
        } else if (sep == " ") {
            return "space"
        } else {
            return sep
        }
    }

    const newline_string = (nl) => {
        if (nl == "\n") {
            return "\\n"
        } else if (nl == "\r\n") {
            return "\\r\\n"
        } else {
            return `${nl}`
        }
    }

    const build_input_test = () => {
        global_data.input_test_count += 1
        let idx = global_data.input_test_count
        let genome = global_data.previous_genome
        let filename = $('#file-upload').prop('files')[0].name
        // defaults
        let primary = global_data.previous_input_action
        let secondary = "gap"
        let title = `${to_title(primary)} ${genome} ${secondary} count`

        let d = document.getElementById("content")

        d.insertAdjacentHTML(
            "beforeend",
            `<div class="container-fluid row pb-2">
                <div class="d-flex col-1 text-primary justify-content-end">
                    <code>In [${idx}]</code>
                </div>
                <div class="col-lg-11 col-sm-12 bg-light border p-2" id="hypothesis-${idx}-anchor">
                    <div class="row">
                        <div class="col">
                            <div class="form-floating">
                                <select class="form-select hyp-test" id="hypothesis-primary-${idx}">
                                    <option selected>Calculate interval densities</option>
                                    <option>Calculate interval counts</option>
                                    <option value="Intersect">Overlap intervals with track</option>
                                </select>
                                <label for="hypothesis-primary-${idx}">Action</label>
                            </div>
                        </div>
                        <div class="col">
                            <div class="form-floating">
                                <select class="form-select hyp-test" id="hypothesis-genome-${idx}">
                                    <option>hg38</option>
                                    <option>hg19</option>
                                </select>
                                <label for="hypothesis-genome-${idx}">Genome build</label>
                            </div>
                        </div>
                        <div class="col" id="hypothesis-secondary-wrapper-${idx}" hidden>
                            <div class="form-floating">
                                <select class="form-select hyp-test" id="hypothesis-secondary-${idx}">
                                    <option selected>gap</option>
                                </select>
                                <label for="hypothesis-secondary-${idx}">Table</label>
                            </div>
                        </div>
                        <div class="col-1">
                            <div class="d-grid gap-2">
                                <button type="button" class="btn btn-primary btn-floating position-relative hyp-test" id="hypothesis-select-action-${idx}">
                                    Run
                                </button>
                            </div>
                        </div>
                    </div>
                    <div class="row">
                        <div class="col-12 py-2 text-secondary" id="hypothesis-blurb-${idx}">
                        </div>
                    </div>
                </div>
            </div>
            `
        )

        $(`#hypothesis-primary-${idx}`).val(global_data.previous_input_action)
        $(`#hypothesis-genome-${idx}`).val(genome);

        update_hypothesis_blurb();

        if (!(global_data.previous_input_action.startsWith("Calculate"))) {
            // show all selects
            $(`#hypothesis-secondary-wrapper-${idx}`).prop("hidden", false)
        }

        // 'Run' click
        document.getElementById(`hypothesis-select-action-${idx}`).addEventListener("click", run_hypothesis_action);
        // select input changes
        let elements = document.getElementsByClassName("form-select hyp-test");
        Array.from(elements).forEach(function (element) {
            element.addEventListener('change', update_hypothesis_blurb);
        });

        return Promise.resolve()
    }

    const run_hypothesis_action = async () => {
        let idx = global_data.input_test_count
        let action = $(`#hypothesis-primary-${idx}`).val()
        let genome = $(`#hypothesis-genome-${idx}`).val();
        let track = $(`#hypothesis-secondary-${idx}`).val();

        global_data.previous_input_action = action
        global_data.previous_genome = genome

        // disable the components of the current test
        $(".hyp-test:enabled").prop("disabled", true)

        if (action == "Calculate interval densities") {
            let chromosome_lengths;
            jQuery.getJSON(`https://api.genome.ucsc.edu/list/chromosomes?genome=${genome}`, (d) => {
                chromosome_lengths = d.chromosomes;
                // adds non-chr prefixed chroms in addition to chr prefixed result
                Object.keys(chromosome_lengths).map(k => chromosome_lengths[k.substring(3)] = chromosome_lengths[k]);
            }).then(() => {
                let genome_length = 0;
                global_data.chromosomes.filter(x => x in chromosome_lengths).map(x => genome_length += chromosome_lengths[x]);

                let d = document.getElementById("content")
                d.insertAdjacentHTML(
                    'beforeend',
                    `<div class="container-fluid row pb-3">
                        <div class="d-flex col-1 justify-content-end"><code>Out[${idx}]</code></div>
                        <div class="col-lg-11 p-0">
                            <div class="d-flex flex-row">
                                <div class="pe-2">
                                    <h6>Interval density across ${genome}</h6>
                                </div>
                                <a class="btn btn-primary btn-circle" data-bs-toggle="offcanvas" href="#offcanvas-interval-density" role="button" aria-controls="offcanvas-interval-density">?</a>
                            </div>
                            <div class="row pt-1">
                                <div class="col-12">
                                    <div id="hypothesis-${idx}-plot"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                    `
                )

                build_bar_plot(
                    global_data.chromosomes,
                    global_data.chromosomes.map(x => global_data.chr_interval_count[x] / global_data.interval_count_filtered),
                    "Chromosome",
                    "Density",
                    `hypothesis-${idx}-plot`,
                    {
                        x: global_data.chromosomes,
                        y: global_data.chromosomes.map(x => (chromosome_lengths[x] / genome_length)),
                        type: "scatter",
                        name: "Expected",
                        mode: "markers",
                        marker: {
                            size: 12, symbol: "circle", color: "#f28e2c", line: {
                                color: 'rgb(255,255,255)',
                                width: 1
                            },
                        }
                    }
                );
                build_input_test()
            })
        } else if (action == "Calculate interval counts") {
            let chromosome_lengths;
            jQuery.getJSON(`https://api.genome.ucsc.edu/list/chromosomes?genome=${genome}`, (d) => {
                chromosome_lengths = d.chromosomes;
                // adds non-chr prefixed chroms in addition to chr prefixed result
                Object.keys(chromosome_lengths).map(k => chromosome_lengths[k.substring(3)] = chromosome_lengths[k]);
            }).then(() => {
                let genome_length = 0;
                global_data.chromosomes.filter(x => x in chromosome_lengths).map(x => genome_length += chromosome_lengths[x]);

                let d = document.getElementById("content")
                d.insertAdjacentHTML(
                    'beforeend',
                    `<div class="container-fluid row pb-3">
                        <div class="d-flex col-1 justify-content-end"><code>Out[${idx}]</code></div>
                        <div class="col-lg-11 p-0">
                            <div class="d-flex flex-row">
                                <div class="pe-2">
                                    <h6>Interval counts across ${genome}</h6>
                                </div>
                                <a class="btn btn-primary btn-circle" data-bs-toggle="offcanvas" href="#offcanvas-interval-counts" role="button" aria-controls="offcanvas-interval-counts">?</a>
                            </div>
                            <div class="row pt-1">
                                <div class="col-12">
                                    <div id="hypothesis-${idx}-plot"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                    `
                )

                // only plot expected for chroms matched to UCSC
                build_bar_plot(
                    global_data.chromosomes,
                    global_data.chromosomes.map(x => global_data.chr_interval_count[x]),
                    "Chromosome",
                    "Count",
                    `hypothesis-${idx}-plot`,
                    {
                        x: global_data.chromosomes,
                        y: global_data.chromosomes.map(y => global_data.interval_count_filtered * (chromosome_lengths[y] / genome_length)),
                        type: "scatter",
                        name: "Expected",
                        mode: "markers",
                        marker: {
                            size: 12, symbol: "circle", color: "#f28e2c", line: {
                                color: 'rgb(255,255,255)',
                                width: 1
                            },
                        }
                    }
                );
                build_input_test()
            })
        } else if (action == "Intersect") {
            await bedtools_intersect(genome, track, global_data.chromosomes, global_data.chr_prefix)
            build_input_test()
        } else {
            console.log("Unaccounted for action selected.")
        }
        scroll_to(`hypothesis-${idx}-anchor`)
    }

    const update_hypothesis_blurb = (event) => {
        let idx = global_data.input_test_count
        let filename = $('#file-upload').prop('files')[0].name
        let primary = $(`#hypothesis-primary-${idx}`).val()
        let secondary = $(`#hypothesis-secondary-${idx}`).val()
        let genome = $(`#hypothesis-genome-${idx}`).val();
        let str = "";

        if (primary.startsWith("Calculate")) {
            str = `Select the genome build of ${filename}.`
            // hide the secondary select box
            $(`#hypothesis-secondary-wrapper-${idx}`).prop("hidden", true)
        } else if (primary == "Intersect") {
            $(`#hypothesis-secondary-wrapper-${idx}`).prop("hidden", false)
            if (secondary == "gap") {
                str = `If your intervals intersect with many gaps across chromosomes, ${genome} may not be the underlying genome build of ${filename}.`

            }
        }
        $(`#hypothesis-blurb-${idx}`).html(str)
    }

    const update_button = (event) => {
        $("#button-alert").prop("hidden", false)
    }

    let global_data = {
    };

    const file_upload = (event) => {
        $("#introduction").prop("hidden", true)
        // clean up existing divs
        $("#content").empty();

        let d = document.getElementById("content")
        d.insertAdjacentHTML(
            "beforeend",
            `<div class="container-fluid row pb-3">
                <div class="col-lg-8 mx-auto bg-light border" id="file-validation" hidden>
                    <table class="table table-borderless m-0">
                        <tbody id="metadata-table">
                        </tbody>
                    </table>
                </div>
            </div>`
        )

        let bed = [];

        let data = {
            newline: "\n",
            delimiter: "\t",
            n_of_fields: [],
            missing_data: 0,
            interval_count: 0,
            interval_length_total: 0,
            interval_lengths: [],
            chr_interval_length: {},
            filtered_intervals: 0,
            zero_length_intervals: 0,
            extends_beyond_chr: 0,
            chr_prefix: true,
            chromosomes: [],
            all_chromosomes: []
        };
        global_data = {
            input_test_count: 0,
            chr_interval_count: {},
            interval_count_filtered: 0,
            chr_prefix: true,
            previous_input_action: "Calculate interval densities",
            previous_genome: "hg38",
            chromosomes: []
        }
        if ($('#file-upload').prop('files')[0]) {
            Papa.parse($('#file-upload').prop('files')[0], {
                skipEmptyLines: true,
                header: false,
                // delimiter: "\t",
                delimitersToGuess: ["\t", ",", " ", "|", ";", Papa.RECORD_SEP, Papa.UNIT_SEP],
                step: (res, parser) => {
                    data.interval_count += 1;

                    // incorrect line ending
                    if (!(res.meta.linebreak == "\n")) {
                        console.log("aborting due to bad line break")
                        data.newline = res.meta.linebreak
                        // parser.abort()
                    }
                    // not tab delimited
                    if (!(res.meta.delimiter == "\t")) {
                        console.log("aborting due to bad delimiter")
                        data.delimiter = res.meta.delimiter
                        // parser.abort()
                    }
                    // most likely a delimiter issue
                    if (res.data.length < 3) {
                        // this file does not have at least 3 fields per line
                        // offending line is d.interval_count with values
                        // res.data
                        console.log("aborting:", res.data)
                        data.missing_data = data.interval_count
                        parser.abort()
                    }
                    // collect shape of bed data
                    if (!(data.n_of_fields.includes(res.data.length))) {
                        data.n_of_fields.push(res.data.length)
                    }

                    let chr = res.data[0];
                    let current_length = res.data[2] - res.data[1];
                    if (!(chr.startsWith('chr'))) {
                        global_data.chr_prefix = false
                    }

                    // set the chromosome order by appearance in bed file
                    if (!(chr in data.chr_interval_length)) {
                        data.all_chromosomes.push(chr)
                    }

                    if (chr.match(/fix|random|alt|Un|hap/g) === null) {
                        global_data.interval_count_filtered += 1;
                        bed.push(res.data.slice(0, 3))
                        // this eval is duplicated
                        if (!(chr in data.chr_interval_length)) {
                            global_data.chromosomes.push(chr);
                        }
                    }

                    // malformatted entry; 0-based start, 1-based end
                    if (current_length == 0) {
                        data.zero_length_intervals += 1;
                    }

                    data.interval_length_total += current_length;
                    data.interval_lengths.push(current_length);
                    data.chr_interval_length[chr] = (data.chr_interval_length[chr] || 0) + current_length;
                    global_data.chr_interval_count[chr] = (global_data.chr_interval_count[chr] || 0) + 1;
                },
                complete: (res, file) => {
                    // resize the upload area
                    let el = document.getElementById("file-upload")
                    // $(".form-control-lg::file-selector-button").attr("style", "min-height:")
                    // show the file stats
                    $("#file-validation").prop("hidden", false)
                    if (!(data.delimiter == "\t")) {
                        add_table_row({ title: "File delimiter", value: delimiter_string(data.delimiter), warn: true, validation: true })
                    } else {
                        add_table_row({ title: "File delimiter", value: delimiter_string(data.delimiter), validation: true })
                    }
                    if (!(data.newline == "\n")) {
                        add_table_row({ title: "File newline", value: newline_string(data.newline), warn: true, validation: true })
                    } else {
                        add_table_row({ title: "File newline", value: newline_string(data.newline), validation: true })
                    }
                    if (!(data.missing_data == 0)) {
                        add_table_row({ title: "Incomplete data on line", value: data.missing_data, warn: true, validation: true })
                        return
                    } else {
                        // < !--0 - length intervals-- >
                        if (data.zero_length_intervals > 0) {
                            add_table_row({ title: "0-length intervals", value: data.zero_length_intervals.toLocaleString(), warn: true, validation: true })
                        } else {
                            add_table_row({ title: "0-length intervals", value: data.zero_length_intervals.toLocaleString(), validation: true })
                        }
                        // < !--number of fields per line-- >
                        add_table_row({ title: "BED fields", value: data.n_of_fields.join(",") })
                        // < !--interval count-- >
                        add_table_row({ title: "Total interval count", value: data.interval_count.toLocaleString() })
                        // < !--total nucleotides-- >
                        add_table_row({ title: "Total nucleotides", value: data.interval_length_total.toLocaleString() })
                        // show the interval size dist
                        add_plot_div('interval-size-dist', 'Interval size distribution', 5).then(() => {
                            var trace = {
                                x: data.interval_lengths,
                                type: 'histogram',
                                marker: { color: aesthetics.palette[0], },
                            };
                            let layout = {
                                height: aesthetics.height,
                                font: aesthetics.labelfont,
                                padding: { t: 0, r: 0, },
                                margin: { t: 0, r: 0, },
                                xaxis: { title: "Interval length", automargin: true, },
                                yaxis: { title: "Count", automargin: true, },
                                legend: { font: { size: 16 } }
                            };
                            Plotly.newPlot('interval-size-dist-plot', [trace], layout, plot_config)
                        })
                        // sort by start, asc
                        bed.sort(function (a, b) {
                            return parseInt(a[1]) > parseInt(b[1]);
                        });
                        // sort by chrom, asc
                        bed.sort(function (a, b) {
                            return a[0] > b[0];
                        });
                        build_input_test();
                        // convert bed to string
                        let str = "";
                        bed.map(a => str += a.join("\t") + "\n");
                        CLI.fs.writeFile("a.bed", str);

                        global_data.chromosomes.sort(sort_alphanum)
                    }
                }
            })
        }
    }

    const build_bar_plot = (chrs, o, x_title, y_title, div, trace = false) => {
        let traces = [
            {
                x: chrs,
                y: o,
                type: "bar",
                marker: {
                    color: aesthetics.palette[0],
                    line: {
                        color: 'black',
                        width: 1
                    },
                },
                name: "Observed"
            }
        ];

        if (trace) {
            traces.push(trace)
        }

        let layout = {
            height: aesthetics.height,
            font: aesthetics.labelfont,
            padding: {
                t: 0,
                r: 0,
            },
            margin: {
                t: 0,
                r: 0,
            },
            xaxis: {
                title: x_title,
                automargin: true,
                type: "category",
            },
            yaxis: {
                title: y_title,
                automargin: true,
            },
            legend: {
                font: {
                    size: 16
                }
            }
        };

        return Plotly.newPlot(div, traces, layout, plot_config);
    };

    $(document).ready(function () {
        document.getElementById("file-upload").addEventListener('change', file_upload);
    });
</script>

</html>
